mucommander-core/src/main/java/com/mucommander/ui/dialog/pref/general/ShortcutsPanel.java

```diff
@@ -1,259 +1,335 @@
+        panel.add(restoreDefaultButton);
+
+        return panel;
+    }
+
+
+    class DeferredDocumentListener implements DocumentListener {
+
+        private final Timer timer;
+
+        public DeferredDocumentListener(int timeOut, ActionListener listener, boolean repeats) {
+            timer = new Timer(timeOut, listener);
+            timer.setRepeats(repeats);
+        }
+
+        public void start() {
+            timer.start();
+        }
+
+        public void stop() {
+            timer.stop();
+        }
+
+        @Override
+        public void insertUpdate(DocumentEvent e) {
+            timer.restart();
+        }
+
+        @Override
+        public void removeUpdate(DocumentEvent e) {
+            timer.restart();
+        }
+
+        @Override
+        public void changedUpdate(DocumentEvent e) {
+            timer.restart();
+        }
+
+    }
+
+    private JPanel createFilteringPanel() {
+        JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
+        panel.setBorder(BorderFactory.createEmptyBorder());
+
+        final JTextField filterField = new JTextField();
+        filterField.setColumns(10);
+
+        final JComboBox<ActionCategory> combo = new JComboBox<>();
+        combo.addItem(ActionCategory.ALL);
+        for (ActionCategory category : ActionProperties.getActionCategory())
+            combo.addItem(category);
+
+
+        final ActionListener filterActionListener = new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                final ActionCategory selectedActionCategory = (ActionCategory) combo.getSelectedItem();
+
+
+                String f = filterField.getText().toLowerCase().trim();
+                String[] words = f.split("\\s+");
+
+                shortcutsTable.updateModel(new ShortcutsTable.ActionFilter() {
+                    @Override
+                    public boolean accept(String actionId, String rowAsText) {
+                        if (selectedActionCategory != null && !selectedActionCategory.contains(actionId)) {
+                            return false;
+                        }
+                        if (words.length == 0) {
+                            return true;
+                        }
+                        for (String word : words) {
+                            if (!rowAsText.contains(word)) {
+                                return false;
+                            }
+                        }
+                        return true;
+                    }
+                });
+                tooltipBar.showDefaultMessage();
+                if (e.getSource() == filterField) {
+                    filterField.requestFocus();
+                }
+            }
+        };
+
+
+        combo.addActionListener(filterActionListener);
+        combo.setSelectedIndex(0);
+
+        DeferredDocumentListener debounceListener = new DeferredDocumentListener(250, filterActionListener, true);
```

Comment:
I see you pass `true` here to set the time to repeated calls, why could it be repeated? shouldn't the trigger for filtering be when the document listener catches an event?
Also, the event that is dispatched from the event dispatching thread so what's the reason for using the timer? (that may be related to the first question above)

Reply:
Cool, thank you for the feedback...

Well, I'm not a java expert, but afaik the `setRepeats(true)` is only explicitly stating out that the event might be triggered more than once, because i think that the `focusGained` event can be emitted multiple times, even if the focus is already present for that element... 

Since the `filterField` loses focus on an update of the tableData (which was a tough problem to solve), it had to `requestFocus` again on every `actionPerformed`, if the source is `filterField`, so that you can still type after a table update. I'm not very proud of this solution, but it seems to work - a rework of the table update to not change the focus of the current element would have been a cleaner solution.

The timer is used, because if the table is updated on the fly, typing feels stuttering, at least on my system (1. typing, 2. table update, 3. requestFocus again - takes to long to feel natural). 250ms represents the median reaction time of a human, so i think the table update has a pretty good balance between "feels like on-the-fly" and "typing still feels natural".

I don't know, why my IDE thought formatting was a good idea, but if i find the time, i'll do the following things:

- remove the formatting completely - i don't think it is necessary
- check if `timer.setRepeats(true)` is really necessary and remove it if it is not...

After this i'll clean the pull request to only contain the changes required for this feature. Unfortunately it will take some time, because I'm busy at the moment.

Reply:
> Well, I'm not a java expert, but afaik the `setRepeats(true)` is only explicitly stating out that the event might be triggered more than once, because i think that the `focusGained` event can be emitted multiple times, even if the focus is already present for that element...

Ah ok, I see your point.
So yeah, without setting `setRepeats(true)` the timer only triggers one even, even calling `start/stop-start/restart` multiple times.
The downside though is that that we trigger an even every 250msec even when the text within the textfield widget is not changing, but we can live with that (the most important part is to cancel the timer when the focus is lost, as you did).
 
> Since the `filterField` loses focus on an update of the tableData (which was a tough problem to solve), it had to `requestFocus` again on every `actionPerformed`, if the source is `filterField`, so that you can still type after a table update. I'm not very proud of this solution, but it seems to work - a rework of the table update to not change the focus of the current element would have been a cleaner solution.

Interesting, did you find the reason the focus switched? is the table requests the focus?

> The timer is used, because if the table is updated on the fly, typing feels stuttering, at least on my system (1. typing, 2. table update, 3. requestFocus again - takes to long to feel natural). 250ms represents the median reaction time of a human, so i think the table update has a pretty good balance between "feels like on-the-fly" and "typing still feels natural".

Yeah, the resulting user experience is nice indeed.
I wonder if we can replace the time with a single thread as I did once for auto-completion - could you please [take a look at it](https://github.com/mucommander/mucommander/blob/master/mucommander-core/src/main/java/com/mucommander/ui/autocomplete/CompletionType.java#L188) and tell if this can be done similarly?
 
> I don't know, why my IDE thought formatting was a good idea, but if i find the time, i'll do the following things:
> 
> * remove the formatting completely - i don't think it is necessary

Well, the benefit of using a formatter is that styling is more consistent and we reduce/eliminate comment of type "please replace the TAB with 4 whitespaces" in the review. I this it would be better to import [this formatter](https://github.com/mucommander/mucommander/blob/master/mucommander-code-format.xml).

> * check if `timer.setRepeats(true)` is really necessary and remove it if it is not...
> 
> After this i'll clean the pull request to only contain the changes required for this feature. Unfortunately it will take some time, because I'm busy at the moment.

Cool :)

Reply:
I think the things i could to are resolved. See comment below. If you aren't happy with these changes, i can't do anything about it, because my IDE is not able to do the code formatting for me like it is required.

Reply:
@sandreas ok, sorry - I thought this PR is still in the works. thanks for pinging. I'll rebase it soon and apply the formatter